varinfo/generate_umm_var.py

```diff
@@ -0,0 +1,92 @@
+
+
+def generate_collection_umm_var(collection_concept_id: str,
+                                auth_header: str,
+                                cmr_env: CmrEnvType = CMR_UAT,
+                                publish: bool = False) -> UmmVarReturnType:
+    """ Run all the of the functions for downloading and publishing
+        a UMM-Var entry to CMR given:
+
+        * collection_concept_id: Concept ID for collection that variables have
+          been generated for.
+        * cmr_env: URLs for CMR environments (OPS, UAT, and SIT)
+        * auth_header: Authorization HTTP header, containing either:
+          - A LaunchPad token: 'Authorization: <token>'
+          - An EDL bearer token: 'Authorization: Bearer <token>'
+        * publish: Optional argument determining whether to publish the
+          generated UMM-Var records to the indicated CMR instance. Defaults to
+          False.
+
+        Note - if attempting to publish to CMR, a LaunchPad token must be used.
+
+    """
+    granule_response = get_granules(collection_concept_id, cmr_env=cmr_env,
+                                    auth_header=auth_header)
+
+    # Get the data download URL for the most recent granule (NetCDF-4 file)
+    granule_link = get_granule_link(granule_response)
+
+    with TemporaryDirectory() as temp_dir:
+        # Download file to runtime environment
+        local_granule = download_granule(granule_link,
+                                         auth_header,
+                                         out_directory=temp_dir)
+
+        # Parse the granule with VarInfo to map all variables and relations:
+        var_info = VarInfoFromNetCDF4(local_granule)
+
+        # Generate all the UMM-Var records:
+        all_umm_var_records = get_all_umm_var(var_info)
+
+    if publish:
+        # Publish to CMR and construct an output object that is a list of
+        # strings. These strings will be either variable concept IDs or
+        # error messages returned from CMR.
+        publication_response = publish_all_umm_var(collection_concept_id,
+                                                   all_umm_var_records,
+                                                   auth_header, cmr_env)
+
+        return_value = [variable_response
+                        if is_variable_concept_id(variable_response)
+                        else ': '.join([variable_name, variable_response])
+                        for variable_name, variable_response
+                        in publication_response.items()]
+    else:
+        # If not publishing, return the full UMM-Var JSON records
+        return_value = list(all_umm_var_records.values())
+
+    return return_value
+
+
+def is_variable_concept_id(possible_concept_id: str) -> bool:
+    """ A helper function to identify if a given string conforms to the
+        expected structure of a variable concept ID.
+
+    """
+    return bool(re.match(r'^V\d{10}-\w+$', possible_concept_id))
```

Comment:
I think tying this to 10 digits is okay. But at some point CMR will require 11. But that seems like a future issue maybe. Thoughts?

Reply:
Might be a little overkill but we could also do a CMR `VariableQuery` search instead. if that returns true we know it's for sure a variable but I guess this is fine for now.

Reply:
Also do you mind defining `is_variable_concept_id` before using it?

Reply:
😬 will they require 11 or will it be 10 or 11 and how long in the future, is this future owen problem or some new sap?

Reply:
I think we're at something like C2XXXXXXXXX-PROV now, so I think we're talking _years_. Strongly tempted to leave as is and close the can of worms I just opened. (Deal with it Future Owen)

Reply:
> Also do you mind defining is_variable_concept_id before using it?

I understand this better now from your other comment.

I think with this module it makes the most sense to have `generate_collection_umm_var` at the top, because that's the main function in the module that we want an end user to interact with, so it should probably be front and centre (ha - neither of those words are synonyms for "at the top", but you know what I mean).